### DISCUSSION TITLE
Implement a machine learning model to help filter comments

### DISCUSSION THREAD

**Discussion by sbordeyne** [ThioJoe/YT-Spammer-Purge#460] [Ideas]
It's not too complicated to do nowadays. The gist of it is to have a sufficiently large dataset of comments to delete, and account names to train the model on.

The idea would be to use a decision tree such as `CatboostClassifier`, which would IMO work pretty well at recognizing comments, with the right features.

For numerical features, I have a few in mind :
- comment frequency (the number of comments posted per unit of time)
- word frequency (can use something like nltk to tokenize the comment content first)
- word frequency with context (with context of 2, 3, 4, 5 words around each word)
- character sets used in spam comments
- number of URLs in the comment
- ML scoring done on the profile picture (so that offending pictures can be associated with spam -- I'm thinking stuff like nudity)
- creation date of the spam accounts
- length of the comment
- ratio between digits and alphabetical characters
- ratio between unicode and alphabetical characters
- ratio between digits and unicode characters
- ratio between uppercase / lowercase characters
- ratio between punctuation / other characters

Given that this tool is already quite efficient, I believe that implementing machine learning into the tool would help future-proof it, and make it harder for scammers to circumvent the filtering rules in place.

Once the model is trained, scoring is relatively fast, the longest part will be the feature calculation for each comment. There are not that many features to calculate though, and those are pretty straightforward to compute as well.

What are your thoughts on this ?

**Comment by ikkysleepy** [comment]
I think there are a lot more other factors, but are you thinking that ML scoring will replace the pattern recognition ?  Or will the scoring be used to validate the logic?

**Reply by DinhHuy2010** [reply]
you can use tenorflow (so what?)
